x/cron/cron.go

```diff
@@ -0,0 +1,273 @@
+// done using provided marshaler.
+//
+// Always use the same marshaler for ticker and scheduler.
+func NewTicker(h weave.Handler, enc TaskMarshaler) *Ticker {
+	return &Ticker{
+		hn:      h,
+		enc:     enc,
+		results: NewTaskResultBucket(),
+	}
+}
+
+// Ticker allows to execute messages queued for future execution. It does this
+// by implementing weave.Ticker interface.
+type Ticker struct {
+	hn      weave.Handler
+	enc     TaskMarshaler
+	results orm.ModelBucket
+}
+
+var _ weave.Ticker = (*Ticker)(nil)
+
+// Tick implementes weave.Ticker interface.
+//
+// Tick can process any number of messages suitable for execution. All changes
+// are done atomically and apply only on success.
+func (t *Ticker) Tick(ctx context.Context, db store.CacheableKVStore) [][]byte {
+	executed, err := t.tick(ctx, db)
+	if err != nil {
+		// This is a hopeless state. This error is most likely due to a
+		// database issues or some other instance specific problems.
+		// This problem is unique to this instance and this operation
+		// most likely succeeded on other nodes. This means that there
+		// is no way we could continue operating as this instance is
+		// out of sync with the rest of the network.
+		failTask(err)
+	}
+	return executed
+}
+
+// failTask is a variable so that it can be overwritten for tests.
+var failTask = func(err error) {
+	panic(fmt.Sprintf(`
+
+Asynchronous task failed.
+
+This error is most likely due to a database issues or some other instance
+specific problems. This problem is unique to this instance and this operation
+most likely succeeded on other nodes. This means that there is no way we could
+continue operating as this instance is out of sync with the rest of the
+network.
+
+%+v
+
+	`, err))
+}
+
+// tick process any number of tasks. It always returns a response and might
+// return an error. This method is similar to the Tick except it provides an
+// error. This makes it easier for the tests to check the result.
+func (t *Ticker) tick(ctx context.Context, db store.CacheableKVStore) ([][]byte, error) {
+	var executed [][]byte
+
+	now, err := weave.BlockTime(ctx)
+	if err != nil {
+		return executed, errors.Wrap(err, "cannot get current time")
+	}
+
+	for {
+		switch key, raw, err := peek(db, now); {
+		case err == nil:
+			res := TaskResult{
+				Metadata:   &weave.Metadata{Schema: 1},
+				Successful: true,
+			}
+
+			// Each task is processed using its own cache instance
+			// to ensure changes are atomic and task processing
+			// independent.
+			cache := db.CacheWrap()
+
+			auth, msg, err := t.enc.UnmarshalTask(raw)
+			if err != nil {
+				res.Successful = false
+				res.Info = fmt.Sprintf("cannot unmarshal task: %+v", err)
+			} else {
+				taskCtx := withAuth(ctx, auth)
+				tx := &taskTx{msg: msg}
+				if _, err := t.hn.Deliver(taskCtx, cache, tx); err != nil {
+					// Discard any changes that the deliver could
+					// have created. We do not want to persist
+					// those.
+					cache.Discard()
```

Comment:
Okay, you Discard the cache here (which means you shouldn't use it in the future), but it is used in 203 even in this error case.

I would use the cache for the DeliverTx issue and always Discard or Write it (add `else {cache.Write()}`).

All code below can write directly to the underlying DB. If you want the whole ticker to be atomic, then wrap it in the code that calls `tick`. I have to spend more time here, but I think you can clean up caching a bit...

One level for each DeliverTx, which is reverted on failure. But then write results to db even on task failure. Then reflect on how you want to handle the errors if eg. storing the results fail. (Maybe panic is best?)

Reply:
If you just add `Savepoint` Decorator to the cron stack, then this whole cache handling here would be given for free.

Reply:
I have not looked into the cache implementation. If I call `.Discard()` on a cache I can no longer use it?

I think adding cache for the whole function call is not good, as it will cause message execution changes to not be atomic. I need to process each message in its own transaction.

There are two "stages" for each processed message.
1. `Deliver` method is called. If it failed, I do not want to carry further any database change it could do. That is why I discard the cache.
2. Later (line 203) I am persisting the message execution result. I do not want to write it directly into the database as it can fail while the `Deliver` could succeed. 
3. In line 211 I am deleting the scheduled task from the queue.

For each task, I want all changes or nothing. That is why I use cache. Otherwise, it could be possible that for example the `Deliver` succeeded and changed the state but task result was not persisted and task was not removed from the queue.

I think you wrote earlier that certain scenarios are not possible in the current storage implementation due to `panic` that is used. I would prefer to not to use this knowledge as it is not enforced by the interfaces.

Does this make sense?

Reply:
Lines 2 and 3 panic if they return error.
Panic kills the process.

If you want caches... then use two levels... One for the whole function, one sub-cache for Deliver.

> I think adding cache for the whole function call is not good, as it will cause message execution changes to not be atomic. I need to process each message in its own transaction.

Remember nothing is really "atomic". A set of writes to a cache can be written or discarded as a group. A write that fails half-way through will write half the data though.

The lowest level store uses an in-memory view, which is only written to disk on Commit. This is described in CommitKVStore and should be held true for any underlying store implementation,

Reply:
I initially wrote a longer comment, because I do not quite understand what you mean.

Instead of a longer comment, my question is how do you think I should implement this instead? My only wish is to avoid leaking to `x/cron` code information of when the process is terminating.

Reply:
I don't know if this is addressing all concerns. I have made improvements in https://github.com/iov-one/weave/pull/850/commits/57a51fd4c30198c0af7182bbd6c429ae13995620
Please check and let me know if anything can be done better.